SPEC 7: Language adjustments and decorator code generalization (#2)
stefanv/specs: spec-0007/index.md

  - "Stéfan van der Walt <stefanv@berkeley.edu>"
  - "Sebastian Berg <sebastianb@nvidia.com>"
  - "Pamphile Roy <roy.pamphile@gmail.com>"
  - "Matt Haberland <mhaberla@calpoly.edu>"
  - Other participants in the discussion <not.yet@named.org>"
discussion: https://github.com/scipy/scipy/issues/14322
endorsed-by:
---

## Description

Currently, libraries across the ecosystem provide various APIs for seeding random number generation.
This SPEC suggests a unified, pragmatic API, taking into account technical and historical factors.
Adopting such a uniform API will simplify the user experience, especially for those who rely on multiple projects.

> We recommend:

Author: This is not necessarily better, but I ran into a merge conflict, so I resolved it with what I had.

- avoiding the use of global state and legacy bitstream generators, and
- standardizing the usage and interpretation of an `rng` keyword for seeding.

We suggest implementing these principles by:

- deprecating the use of `RandomState` and `numpy.random.seed`,
- accepting an `rng` argument in all functions that require random number generation, and

> - using `numpy.random.default_rng` to validate the `rng` argument and instantiate a `Generator`.

Author: I thought it was worth mentioning accepting an `rng` argument as a separate bullet. It seems like deprecating `random_state`/`seed` is implicit? Do we want to make it an explicit recommendation?

Reviewer: Perhaps good to be explicit, something like: "Deprecate the use of `random_state`/`seed` keywords in favor of an `rng` keyword."

Author: Exactly what I was thinking 👍

### Concepts

> - `BitGenerator`: Generates a stream of pseudo-random bits. The default generator in NumPy (`numpy.random.default_rng`) uses PCG64.

Author: Not sure if this is required or if the link would work either way? If it would work either way, I will revert.

Reviewer: I think the SPEC markdown does not generate links automatically, but this is fine.

### Constraints

NumPy, SciPy, scikit-learn, scikit-image, and NetworkX all implement pseudo-random seeding in slightly different ways.
Common keyword arguments include `random_state` and `seed`.
In practice, the seed is unfortunately also often controlled using `numpy.random.seed`.

## Implementation

Legacy behavior in packages such as scikit-learn (`sklearn.utils.check_random_state`) typically handle `None` (use the global seed state), an int (convert to `RandomState`), or `RandomState` object.

Two strong motivations for moving over to `Generator`s are:

(1) they avoid naïve seeding strategies, such as using successive integers, via the underlying [SeedSequence](https://numpy.org/doc/stable/reference/random/parallel.html#seedsequence-spawning);
(2) they avoid using global state (from `numpy.random.mtrand._rand`).

Our recommendation here is a deprecation strategy which does not in _all_ cases adhere to the Hinsen[^hinsen] principle.

The [deprecation strategy](https://github.com/scientific-python/specs/pull/180#issuecomment-1515248009) is:

> 1. Accept both `rng` and `random_state`/`seed` keyword arguments.

Author: Let's be carefuly to refer to both `random_state` and `seed` together in the text throughout.

Author: ```suggestion
1. Accept both `rng` and `random_state`/`seed` arguments.
```

2. If `rng=None`, handle `random_state`/`seed` as in legacy behavior (see above), except use a compatible Generator instead of RandomState.

>    A DeprecationWarning is raised to warn about a future change in behavior.

Author: Should this section mention all the other things that happen during the deprecation period? The logic is a lot more involved than these four steps suggest.

Reviewer: What do you think is missing here?

Author: See the new documentation of the decorator for all the logic it implements. All of that - whatever we think is important - could be mentioned here.

Author: ```suggestion
2. If both are specified, raise an error.
3. If neither is specified and `np.random.seed` has been used to set the seed, emit a `FutureWarning` about the upcoming change in behavior.
4. If `random_state`/`seed` is passed by keyword or by position, treat it as before, but:
  - Emit a `DeprecationWarning` if passed by keyword, warning about the deprecation of keyword `random_state` in favor of `rng`.
  - Emit a `FutureWarning` if passed by position, warning about the change in behavior of the positional argument.
5. If `rng` is passed by keyword, standardize it using `numpy.random.default_rng`.
```

Reviewer: I'm a tad worried that the instructions are too detailed, causing confusion. Perhaps we can have a broad version, and then a more detailed discussion?

Author: Let's discuss it in real time. I think the implementation has been described in vague terms already, so at some point the details need to be laid out, and probably in English (like this) as well as the code.

Reviewer: OK, that works for me, given that we've already given the high-level overview.

> 3. After <X time>, use only `rng` validated with `numpy.random.default_rng`.

Reviewer: ```suggestion
3. After <X time>, use only `rng`, validated with `numpy.random.default_rng`.
```

  - "Stéfan van der Walt <stefanv@berkeley.edu>"
  - "Sebastian Berg <sebastianb@nvidia.com>"
  - "Pamphile Roy <roy.pamphile@gmail.com>"
  - "Matt Haberland <mhaberla@calpoly.edu>"
  - Other participants in the discussion <not.yet@named.org>"
discussion: https://github.com/scipy/scipy/issues/14322
endorsed-by:
---

## Description

> Currently, libraries across the ecosystem provide various APIs for seeding random number generation.

Author: I think of "seeding" as the process of setting a seed. If that's right, then libraries typically get involved with "seeding" only when the `seed`/`random_state` parameter accepts a seed (e.g. integer) as an argument. 

When the `seed`/`random_state` parameter accepts a `RandomState`/`Generator`, or when the user sets a seed with `np.random.seed`, I don't think the library's API for "seeding" is being exercised.

With that in mind, consider:
```suggestion
Currently, libraries across the ecosystem provide various APIs for controlling random number generation.
```
and using "controlling" in place of "seeding" (and similar) throughout.

Reviewer: I suspect "seeding", while not perfectly accurate, may convey a more intuitive meaning here.

Author: Hmm I thought the opposite, which was the point. But I'll leave this unmerged.

Reviewer: I don't feel strongly about this instance, but the word seeding is already in the SPEC title :grimacing:

Author: K. I'll leave it alone and resolve.

This SPEC suggests a unified, pragmatic API, taking into account technical and historical factors.
Adopting such a uniform API will simplify the user experience, especially for those who rely on multiple projects.

We recommend:

- avoiding the use of global state and legacy bitstream generators, and
- standardizing the usage and interpretation of an `rng` keyword for seeding.

We suggest implementing these principles by:

> - deprecating the use of `numpy.random.seed` to control the random state,

Author: This said something before about deprecating use of `RandomState`s, but that is not really done explicilty. It is implicit: use of old `random_state`/`seed` arguments are deprecated, and the `rng` argument is validated by `default_rng`, so `RandomState`s won't be accepted.

Reviewer: It says it above now, "legacy bitstream generators". We could, in brackets, add `RandomState` there.

> - deprecating the use of `random_state`/`seed` arguments in favor of an `rng` argument in all functions that require random number generation, and

Author: ```suggestion
- deprecating the use of `random_state`/`seed` arguments in favor of a consistent `rng` argument in all functions that require random number generation, and
```
or maybe just

```suggestion
- deprecating the use of `random_state`/`seed` arguments in favor of a consistent `rng` argument, and
```

- using `numpy.random.default_rng` to validate the `rng` argument and instantiate a `Generator`.

> ### Scope

Author: Feel free to refactor. This is my attempt to address the fact that it is not considering other array backends.

This is intended as a recommendation to all libraries that allow users to control the
state of a NumPy random number generator. It is specifically targeted toward functions
that allow use of `numpy.random.seed` to control the random state and accept
`RandomState` instances via an argument with a name other than `rng`, but the ideas
are more broadly applicable. Use of random number generators other than those provided

> by NumPy are not considered at this time, but may be addressed in a future version.

Reviewer: I suspect we won't go into that in the future; this is mainly about saying "use `rng` to set a seed/provide a generator". If projects want to allow foreign generators, that's probably OK, as long as they use the same argument name.

Author: Haha OK, but that would not be strictly compatible with this as written because this specifies that `rng` would be validated with `default_rng`, which doesn't accept foreign Generators.

### Concepts

- `BitGenerator`: Generates a stream of pseudo-random bits. The default generator in NumPy (`numpy.random.default_rng`) uses PCG64.
- `Generator`: Derives pseudo-random numbers from the bits produced by a `BitGenerator`.
- `RandomState`: a [legacy object in NumPy](https://numpy.org/doc/stable/reference/random/index.html), similar to `Generator`, that produces random numbers based on the Mersenne Twister.

### Constraints

NumPy, SciPy, scikit-learn, scikit-image, and NetworkX all implement pseudo-random seeding in slightly different ways.
Common keyword arguments include `random_state` and `seed`.
In practice, the seed is unfortunately also often controlled using `numpy.random.seed`.

## Implementation

Legacy behavior in packages such as scikit-learn (`sklearn.utils.check_random_state`) typically handle `None` (use the global seed state), an int (convert to `RandomState`), or `RandomState` object.

Two strong motivations for moving over to `Generator`s are:

(1) they avoid naïve seeding strategies, such as using successive integers, via the underlying [SeedSequence](https://numpy.org/doc/stable/reference/random/parallel.html#seedsequence-spawning);

> (2) they avoid using global state (from `numpy.random.mtrand._rand`).

Author: Just making sure - you intended for all of this to show up as a paragraph without line breaks, right?

Reviewer: It would probably render better as a numbered list.

Author: I will change that.

Author: Done (locally).

Our recommendation here is a deprecation strategy which does not in _all_ cases adhere to the Hinsen[^hinsen] principle.

The [deprecation strategy](https://github.com/scientific-python/specs/pull/180#issuecomment-1515248009) is:

1. Accept both `rng` and `random_state`/`seed` keyword arguments.
2. If `rng=None`, handle `random_state`/`seed` as in legacy behavior (see above), except use a compatible Generator instead of RandomState.
   A DeprecationWarning is raised to warn about a future change in behavior.

> 3. After <X time>, use only `rng`, validated with `numpy.random.default_rng`.

Author: ```suggestion
6. After the deprecation period, use only `rng`, validated with `numpy.random.default_rng`.
```
Probably understandable as-is. Depends on how formal this is suppose to be.

Reviewer: Should we decide the length of that deprecation period, or is that for the project to determine?

Author: Probably defer to project standards, unless you also want to write a SPEC about that sort of thing! That might be worthwhile, like SPEC 0, to get projects talking about  that, but maybe for a later time? Could specify a minimum length of time, but I think most major projects already do something reasonable. (For SciPy it's two releases, so a year.) But I was thinking it would be unspecified.

   Raise an error if `random_state`/`seed` is provided.

> 4. At a time of the library's choosing, remove any machinery related to `random_state`/`seed`.

Author: After the deprecation period, any machinery related to `random_state`/`seed` is no longer used, so I think it's implicit that it can be removed.
```suggestion
```
In any case, this goes a bit beyond the deprecation strategy, and this section is longer than it was, so might as well try this.

  - "Stéfan van der Walt <stefanv@berkeley.edu>"
  - "Sebastian Berg <sebastianb@nvidia.com>"
  - "Pamphile Roy <roy.pamphile@gmail.com>"
  - "Matt Haberland <mhaberla@calpoly.edu>"
  - Other participants in the discussion <not.yet@named.org>"
discussion: https://github.com/scipy/scipy/issues/14322
endorsed-by:
---

## Description

Currently, libraries across the ecosystem provide various APIs for seeding random number generation.
This SPEC suggests a unified, pragmatic API, taking into account technical and historical factors.
Adopting such a uniform API will simplify the user experience, especially for those who rely on multiple projects.

We recommend:

- avoiding the use of global state and legacy bitstream generators, and
- standardizing the usage and interpretation of an `rng` keyword for seeding.

We suggest implementing these principles by:

- deprecating the use of `numpy.random.seed` to control the random state,
- deprecating the use of `random_state`/`seed` arguments in favor of a consistent `rng` argument, and

> - using `numpy.random.default_rng` to validate the `rng` argument and instantiate a `Generator`.

Author: Note that `RandomState` is not a valid value "default_rng does not accept a `RandomState` as an argument or use the global random seed."

Author: After the bulleted list, added:

> Note that `numpy.random.default_rng` does not accept instances of `RandomState`, so use of `RandomState` to control the seed is effectively deprecated, too.

### Scope

This is intended as a recommendation to all libraries that allow users to control the
state of a NumPy random number generator. It is specifically targeted toward functions
that allow use of `numpy.random.seed` to control the random state and accept
`RandomState` instances via an argument with a name other than `rng`, but the ideas
are more broadly applicable. Use of random number generators other than those provided

> by NumPy are not considered at this time, but may be addressed in a future version.

Author: ```suggestion
by NumPy is beyond the scope of this SPEC.
```

Author: Done (locally)

   ```

   Also note the suggested language for the `rng` parameter docstring, which encourages the user to pass a `Generator` or None, but allows for other types accepted by `numpy.random.default_rng`.

### Impact

The following users will be affected:

1. Those who use `numpy.random.seed`. 
   The proposal will do away with that global seeding mechanism, meaning that code that relies on it will, after a certain deprecation period, start seeing a different stream of random numbers than before.
   To ensure that this does not go unnoticed, the library should raise a `FutureWarning` if `np.random.seed` has been called. (See Code below for an example.)

>    Such code will, in effect, go from being seeded to being unseeded. 

Author: Merge conflict. If removing "Such" was intentional, we can do that.

Reviewer: I wasn't sure what the "Such" referred to, so took it out.

Author: OK, adjusted this part a bit more.

   To prevent this, users must switch from using `numpy.random.seed` to passing the `rng` argument explicitly to all functions that accept it.

2. Those who do not seed. Their code will, after the deprecation period, use the newly proposed default. Since they were already not requesting repeatable sequences, and since the underlying _distributions_ of pseudo-random numbers did not change, they should be unaffected.

3. Users of `random_state`/`seed` arguments. 
   Support for these arguments will be dropped eventually, but during the deprecation period, we can provide clear guidance, via warnings and documentation, on how to migrate to the new `rng` keyword.

[^hinsen]: The Hinsen principle states, loosely, that code should, whether executed now or in the future, return the same result, or raise an error.

### Code

> As an example, consider how a SciPy function would transition from a `seed` parameter to the `rng` keyword using a decorator.

Author: ```suggestion
As an example, consider how a SciPy function would transition from a `random_state` parameter to the `rng` keyword using a decorator.
```

   ```

   Also note the suggested language for the `rng` parameter docstring, which encourages the user to pass a `Generator` or None, but allows for other types accepted by `numpy.random.default_rng`.

### Impact

There are three classes of users, which will be affected to varying degrees.

1. Those who do not attempt to control the random state.
   Their code will immediately switch from using the unseeded global `RandomState` to using an unseeded `Generator`.

>    Since the underlying _distributions_ of pseudo-random numbers will not change, these users will be relatively unaffected.

Reviewer: I wonder about that word "relatively". Maybe "should be unaffected"?

Author: OK.

Author: Done.

   ```

   Also note the suggested language for the `rng` parameter docstring, which encourages the user to pass a `Generator` or None, but allows for other types accepted by `numpy.random.default_rng`.

### Impact

> There are three classes of users, which will be affected to varying degrees.

Author: I changed this wording because one of the classes of users is not affected.

1. Those who do not attempt to control the random state.

>    Their code will immediately switch from using the unseeded global `RandomState` to using an unseeded `Generator`.

Author: Note I've reordered these in order of least to greatest impact.
Previously this one said "Their code will, after the deprecation period, use the newly proposed default." I don't think that's true, given the behavior fo `check_random_state`. They *immediately* begin to use `default_rng` instead of the unseeded global `RandomState`. Right?

   Since the underlying _distributions_ of pseudo-random numbers will not change, these users will be relatively unaffected.

2. Users of `random_state`/`seed` arguments. 
   Support for these arguments will be dropped eventually, but during the deprecation period, we can provide clear guidance, via warnings and documentation, on how to migrate to the new `rng` keyword.

3. Those who use `numpy.random.seed`. 

>    The proposal will do away with that global seeding mechanism, meaning that code that relies on it will, after the deprecation period, go from being seeded to being unseeded.

Author: ```suggestion
   The proposal will do away with that global seeding mechanism, meaning that code that relies on it would, after the deprecation period, go from being seeded to being unseeded.
```

>    To ensure that this does not go unnoticed, libraries that allowed for control of the random state via `numpy.random.seed` should raise a `FutureWarning` if `np.random.seed` has been called. (See Code below for an example.)

Author: ```suggestion
   To ensure that this does not go unnoticed, libraries that allowed for control of the random state via `numpy.random.seed` should raise a `FutureWarning` when using the global random state after `np.random.seed` has been called. (See Code below for an example of how to check.)
```

Reviewer: I'm not sure this is right; isn't the intention to warn in two cases: when using `rng` along with global seeding, or when using the global seed when `random_state=None`? So, I think any time that we detect that `np.random.seed` has been called, we need to warn the user that they're doing something wrong.

Author: > when using rng along with global seeding

How can `rng` be used with global seeding? Arguments to `rng` are validated by `default_rng`, and that will never return the global `RandomState` (or any `RandomState`).

> when using the global seed when `random_state=None`

Yes, currently when `random_state=None` *because it has not been specified* and `np.random.seed` has been called, we give a `FutureWarning` that the global random state won't work in the future.

When `random_state` *is specified explicitly*, whether it is `None`, `np.random`, or something else, we need to give a warning about `random_state` being used. If it was used by keyword, that is deprecated. It it was used positionally, there's a `FutureWarning` about that behavior changing. Either way, the user needs to use `rng` with a valid input to avoid the warning, and `rng` doesn't permit use of the global random state, so I don't think we *also" need to give a warning against use of the global random state. But we can discuss it.

Reviewer: I meant by that: calling seed and then also specifying an rng. Sorry for being unclear.

Reviewer: Perhaps something like: libraries should raise a FutureWarning when the `random_state` argument is set in a way that accesses global random state (this will not longer be supported via `rng`).

Author: Feel free to change it. To avoid more merge conflicts, I will not be working on this PR locally.

   In response, users must switch from using `numpy.random.seed` to passing the `rng` argument explicitly to all functions that accept it.

[^hinsen]: The Hinsen principle states, loosely, that code should, whether executed now or in the future, return the same result, or raise an error.

> ### Code

Author: To resolve https://github.com/stefanv/specs/pull/2/files#r1677000571, we could include in the SPEC only the code for checking whether `np.random.seed` has been used, and perhaps afterwards we could link to implementations by projects? I think it was a good exercise to write the code, but in the interest of getting the SPEC done, I think it should be left out, and we can link to it after it's been implemented by a project (so others can copy if desired).

    - If the function is called with both `random_state` and `rng`, the decorator
      raises an error.
    - If `random_state` is provided as a keyword argument, the decorator passes
      `random_state` to the function's `rng` argument as a keyword, but the decorator
      will emit a `DeprecationWarning` about the deprecation of keyword `random_state`.
    - If `random_state` is provided as a positional argument, the decorator passes
      `random_state` to the function's `rng` argument by position, but the decorator
      will emit a `FutureWarning` about the changing behavior of the argument.
    - If `rng` is provided as a keyword argument, the decorator validates `rng` using
      `numpy.random.default_rng` before passing it to the function.
    - If neither `random_state` nor `rng` is provided, the decorator checks whether
      `np.random.seed` has been used to set the global seed. If so, it emits a
      `FutureWarning` against use of `numpy.random.seed`. Either way, the decorator
      calls the function without explicitly passing the `rng` argument.

    To avoid warnings, a user must pass `rng` as a keyword or pass a `Generator`

>     object or `None` by position.

Reviewer: We discussed, but I still wonder if we aren't being too lenient. If the library lets this go past, the *same code* will produce very different results in old and new versions?

If they error on positional None, then at least they *know* when None is sent to mean "use system entropy" vs "use global seed state".

Author: I realized this was no good and took out this behavior a while ago. It will always warn with positional input now. I'll change this part of the text.

                               f"deprecated and replaced by `{new_name}`.  "
                               f"Support for `{old_name}` will be removed "
                               f"in SciPy {end_version}. Use keyword `{new_name}` "
                               "to silence this warning.")
                    warnings.warn(message, DeprecationWarning, stacklevel=2)

            elif as_pos_arg and dep_version:
                # Warn about changing meaning of positional arg

                # Note that this decorator does not deprecate positional use of the
                # argument; it only warns that the behavior will change in the future.
                # Simultaneously transitioning to keyword-only use is another option.

                message = (f"Positional use of `{new_name}` (formerly known as "
                           f"`{old_name}`) is still allowed, but the behavior is "
                           "changing: the argument will be validated using "
                           "`np.random.default_rng` beginning in SciPy {end_version} "
                           f"Ensure that `np.random.default_rng(rng)` returns "

>                            "successfully to silence this warning.")

Author: Add short list of gotchas to this message.
> Note that the behavior of `default_rng` is different for several common inputs:
> None: unpredictable random numbers
> np.random: will raise an error
> integer: will get a different stream of random numbers
>`RandomState` instance: will raise an error

   It changes the documentation and auto-completion to only advertise the new `rng` parameter.

```python
import numpy as np
import functools
import numbers
import warnings

_NoValue = object()  # singleton to indicate not explicitly passed

> # Should this specify the version number in which the behavior will change?

Author: ```suggestion
```
Should the warning specify the version number in which the behavior will change?

Reviewer: :shrug:

Author: Let's address this in the main PR after this PR is merged. But in SciPy, if we emit a `DeprecationWarning` or `FutureWarning`, it now specifies the version in which the behavior will change. The clear timeline is not just important to users, but also developers. (In the past, ambiguous deprecations often were left as warnings for several years.)

Author: Added version number.

def _check_random_state(seed=_NoValue, rng=_NoValue):
    """ PRNG validator during deprecation period
    
    Accepts inputs specified via old `seed` (or `random_state`) arguments

>     or new `rng` argument and returns appropriate PRNG.

Reviewer: ```suggestion
    as well as new `rng` argument and returns appropriate PRNG.
```

Author: I suppose that's correct in that it accepts them both and raises an error if both are present 👍 . I replaced "or" with "and".

Author: Yeah, this must have been in `check_random_state`, which is gone. Resolving.

>     Once the removal of `seed`/`random_state` argument is executed,

Reviewer: ```suggestion
    Once the removal of `seed`/`random_state` argument is enacted,
```

Author: I think the definition of executed that is "carry out or put into effect (a plan, order, or course of action)" is ideal here, but I could compromise on "complete". Of course, you are also welcome to change it afterwards.

Reviewer: Yes, I think technically you are correct with "executed", just that it has an overloaded connotation when it comes to computer software. "completed" sounds good to me.

Author: I'll change it to "completed".

Author: The text has changed and this no longer seems to apply; "executed" only appears in the description of the Hinsen principle. Resolving.

    this function may be removed, and `numpy.random.default_rng` may be
    used to validate the `rng` argument.
    
    Parameters
    ----------

>     seed : <insert long list of objects here>

Reviewer: Maybe this just becomes "any", and the next one becomes `Generator, int, optional` or something like that.

Author: I figured I'd let you fix this part because you have been working on the types.

Author: `check_random_state` is gone; resolving!

        Input specified via old `seed` (or `random_state`) argument
    rng :  <insert long list of objects here>
        Input specified via old `seed` (or `random_state`) argument
    Returns
    -------
    prng : NumPy `Generator` or `RandomState`
        Pseudo-random number generator
    """
    if rng is not _NoValue and seed is not _NoValue:

>         raise TypeError("Cannot pass both `rng` and `seed` (or `random_state`).")

Author: ```suggestion
        raise TypeError("Cannot pass both `rng` and `random_state`/`seed`.")
```

        return seed

    raise ValueError(f"'{seed}' cannot be used to seed a numpy.random.RandomState"
                     " instance.")

def _prepare_rng(old_name, position_num=None, dep_version=None):
    """ Example decorator to deprecate old PRNG usage

> 

Reviewer: Maybe we need a paragraph here to describe what it does, given that very few people are likely to read the code.

Author: Added. Please check me.
You asked yesterday whether I thought it was a good thing to include the code. I said yes then because it helps projects, but another benefit is that it forces us to think through the implications of the SPEC very carefully.

                # Not required for adoption of this SPEC; feel free to remove
                args = list(args)
                args[position_num] = _check_random_state(seed=args[position_num])
                if dep_version:
                    message = (f"Positional use of argument `{old_name}` and "
                               "those that follow is deprecated and will cause "
                               f"an error in SciPy {end_version}. Replace these "
                               f"with keyword arguments, and use `{new_name}` "
                               f"instead of `{old_name}`.")
                    warnings.warn(message, DeprecationWarning, stacklevel=2)

            elif as_new_kwarg:  # no warnings; this is the preferred use
                kwargs[new_name] = _check_random_state(rng=kwargs[new_name])

            else:  # will emit FutureWarning if `np.random.seed` was used
                kwargs[new_name] = _check_random_state()

            return fun(*args, **kwargs)
        return wrapper
    return decorator

# Example usage of _prepare_rng decorator
@_prepare_rng("random_state", 1, "1.15.0")

> def library_function(arg1, rng=None, arg2=None):

Author: Note: during the deprecation period it will still look like `rng` can be specified as a positional argument.

Author: ~~Actually, I think I can change that. I'll do that tonight.~~
I could, but not without the user specifying the names of the remaining arguments or doing introspection.

This brings up https://github.com/scientific-python/specs/pull/180/files#r1662814317 .

Currently, we're in this halfway state where the signature will look neither like it did before nor what it is going to be. The `random_state` parameter has been renamed, but `rng` and following positional arguments are not shown as keyword-only like they will become. That is not ideal. We really should be doing one of these two things during the deprecation:

1. Leave the signature unchanged except for adding a keyword-only `rng` argument. The advantage is that we can show a deprecation admonition for the `random_state` argument in the documentation. The added complexity compared to the current code is that we may want the decorator to inject the admonition using [`_docscrape`](https://github.com/scipy/scipy/blob/main/scipy/_lib/_docscrape.py) (which SciPy, at least, already vendors). Not complicated - it would just take a few lines. For example, [here](https://github.com/scipy/scipy/blob/d48a13af42c10145b91031ba7a737ac9a439c29e/scipy/stats/_axis_nan_policy.py#L652-L671) is some code that uses docscrape `FunctionDoc` to standardize the documentation of several parameters of `scipy.stats` functions.
2. Make the signature as it is going to be after the deprecation, with `random_state` replaced by a keyword-only `rng`, and all positional/keyword arguments that came after `random_state` made keyword-only, too. The advantage is that it makes the future signature very clear. The added complexity is that the decorator has to pass all the remaining parameters to the function as keyword-only, so either the user has to provide their names or they need to be determined by introspection. Also not hard, but cumbersome and somewhat confusing to make the user of the decorator pass the names of all the arguments, and some people like say that `inspect` is too slow and complex to be used (even if it is much faster and simpler than the rest of the code).

[This issue](https://github.com/stefanv/specs/pull/2#discussion_r1676973078) may force us to go with option 1. `_check_random_state` assumes that it knows how the `random_state` parameter was treated before the deprecation and converts it to a `RandomState` object, but the truth is that we don't actually know what libraries were doing with `random_state` before. So if we want to be possible for libraries to use the decorator as-is without immediate backward compatibility concerns, we can warn based on what we get for `random_state`, but we really should just pass it through to the `random_state` parameter of the original function without modifying it. (Incidentally, this would simplify `_check_random_state` quite a bit or eliminate the need for it entirely.)

It's really your call, @stefanv! LMK if you'd like to discuss in real time.

Author: OTOH, there seem to be some opinions against including it in the SPEC. I don't imagine anyone would complain if we provide it separately from the SPEC, though, so I'll take it out. (We still need to make this decision but it doesn't have to hold up the SPEC merging.)

Author: Resolved. `rng` will immediately appear in the signature. No longer deprecating positional use of argument.

    # The decorated library function accepts an `rng` argument, but the user
    # can also pass a `random_state` argument (as specified by `old_name`) or
    # a positional argument (with position specified by `position_name`). In
    # any case, the argument passed to `library_function` will be either a

>     # Generator or a RandomState after validation by the decorator.

Author: We should specify what happens *after* the deprecation period is over.
If positional use of the PRNG argument is deprecated, then *rng* (and all following arguments) are marked keyword-only.
The decorator gets removed, and `rng = np.random.default_rng(rng)` can be used to validate the argument inside the function?

    assert isinstance(rng, (np.random.Generator, np.random.RandomState))
    print((arg1, rng.random(), arg2))
```

### Core Project Endorsement

Endorsement of this SPEC means that a project intends to:

- avoid the use of global state and legacy bitstream generators, and
- standardize the usage and interpretation of an `rng` keyword for seeding.

### Ecosystem Adoption

> To adot this SPEC, a project should:

Author: ```suggestion
To adopt this SPEC, a project should:
```

- deprecate the use of `RandomState` and `numpy.random.seed`,

> - accept an `rng` argument in all functions that require random number generation, and

Author: Again - specify that `random_state`/`seed` arguments should be deprecated?

        return seed

    raise ValueError(f"'{seed}' cannot be used to seed a numpy.random.RandomState"
                     " instance.")

def _prepare_rng(old_name, position_num=None, dep_version=None):
    """ Example decorator to deprecate old PRNG usage

    Suppose the decorator is applied to a function that used to accept parameter
    `old_name='random_state'` either by keyword or as a positional argument at
    `position_name=1`. At the time of application, the name of the argument in the
    function signature is manually changed to the new name, `rng`.
    - If the function is called with both `random_state` and `rng`, the decorator
      raises an error.
    - If `random_state` is provided as a keyword argument, the decorator validates

>       the argument according to the old rules* and passes it to the function as `rng`,

Author: Oops, I meant to specify what `_check_random_state` assumes the "old rules" to be:
- `None` becomes the global `RandomState`.
- integers seed a new `RandomState`
- `RandomState` or `Generator` instances are passed through unchanged
- Everything else is rejected.

If this is not what the wrapped function used to do, the decorator/`_check_random_state` cannot be applied as-is, so I think it's important to document this.

Where should this list go?

Author: No longer needed.
